File: mobile/lib/data/models/user_role.dart
enum UserRole {
  superAdmin('SUPER_ADMIN', '超级管理员', '可管理用户、角色与全部后台能力'),
  librarian('LIBRARIAN', '馆员', '可管理图书、批注与扫描任务'),
  reader('READER', '读者', '仅使用阅读与同步功能');

  const UserRole(this.value, this.label, this.description);

  final String value;
  final String label;
  final String description;

  bool get canAccessAdmin => this == superAdmin || this == librarian;

  bool get canManageAdminUsers => this == superAdmin;

  static UserRole fromValue(String? value) {
    final normalized = value?.trim().toUpperCase();
    for (final role in values) {
      if (role.value == normalized) {
        return role;
      }
    }
    return reader;
  }
}

File: backend/app/src/main/kotlin/com/privatereader/auth/UserRole.kt
package com.privatereader.auth

enum class UserRole(val value: String) {
    SUPER_ADMIN(RoleNames.SUPER_ADMIN),
    LIBRARIAN(RoleNames.LIBRARIAN),
    READER(RoleNames.READER),
    ;

    companion object {
        val allValues: Set<String> = entries.mapTo(linkedSetOf()) { it.value }
        val adminAccessValues: List<String> = listOf(SUPER_ADMIN.value, LIBRARIAN.value)

        fun normalize(raw: String): String {
            val normalized = raw.trim().uppercase()
            require(normalized in allValues) { "Unsupported role" }
            return normalized
        }

        fun hasGlobalLibraryAccess(role: String): Boolean = role in adminAccessValues
    }
}

object RoleNames {
    const val SUPER_ADMIN = "SUPER_ADMIN"
    const val LIBRARIAN = "LIBRARIAN"
    const val READER = "READER"
}

object RoleAuthorities {
    const val PREFIX = "ROLE_"
    const val SUPER_ADMIN = PREFIX + RoleNames.SUPER_ADMIN
    const val LIBRARIAN = PREFIX + RoleNames.LIBRARIAN
    const val READER = PREFIX + RoleNames.READER

    fun fromRole(role: String): String = PREFIX + role
}

object RoleExpressions {
    const val SUPER_ADMIN_ONLY = "hasAuthority('" + RoleAuthorities.SUPER_ADMIN + "')"
    const val ADMIN_ACCESS =
        "hasAnyAuthority('" + RoleAuthorities.SUPER_ADMIN + "','" + RoleAuthorities.LIBRARIAN + "')"
}
